src/syn/io/der.rs

```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+        }
     }
 
     #[inline]
     fn read_ia5string<C: ia5string::Constraint>(&mut self) -> Result<String, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[ia5string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 22 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        if let Length::Definite(l) = length {
+            let octets = self.buffer.read_octet_string(l)?;
+            String::from_utf8(octets).map_err(|_| Self::Error::InvalidUtf8String)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_octet_string<C: octetstring::Constraint>(&mut self) -> Result<Vec<u8>, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[octet string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 4 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        if let Length::Definite(l) = length {
+            self.buffer.read_octet_string(l)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_bit_string<C: bitstring::Constraint>(&mut self) -> Result<(Vec<u8>, u64), Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[bit string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Length::Definite(l) = length {
+            let octets = self.buffer.read_octet_string(l)?;
+            let size = (&octets.len() * 8) as u64;
+            Ok((octets, size))
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_boolean<C: boolean::Constraint>(&mut self) -> Result<bool, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[boolean] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let (PC::Primitive, 1, Length::Definite(1)) = (&pc, &tag, &length) {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```

Comment:
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 1) || !matches!(len, LengthDefinitive(1)) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_BOOLEAN != tag || !matches!(length, Length::Definitive(1)) {
    return Err(Error::InvalidType);
}
```